core/client/src/client.rs

```diff
@@ -1343,6 +1350,20 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 	}
 }
 
+impl<B, E, Block, RA> ProvideUncles<Block> for Client<B, E, Block, RA> where
+	B: backend::Backend<Block, Blake2Hasher>,
+	E: CallExecutor<Block, Blake2Hasher>,
+	Block: BlockT<Hash=H256>,
+{
+	fn uncles(&self, target_hash: Block::Hash, max_generation: NumberFor<Block>) -> error::Result<Vec<Block::Header>> {
+		Client::uncles(self, target_hash, max_generation)?
+			.into_iter()
+			.map(|hash| Client::header(self, &BlockId::Hash(hash))
+				.map(|maybe_header| maybe_header.expect("Each uncle exists in the database; qed")))
```

Comment:
it would be much safer to `filter_map`. this expectation doesn't prove to me why the uncle will exist (in the future, it might race against block pruning or something similar)